linkerd/app/outbound/src/resolve.rs

```diff
@@ -0,0 +1,118 @@
+#![allow(warnings)]
+
+use crate::endpoint;
+use futures::{future, prelude::*, stream};
+use linkerd2_app_core::{
+    is_discovery_rejected,
+    proxy::{
+        api_resolve::Metadata,
+        core::{Resolve, ResolveService, Update},
+        discover::{self, Buffer, FromResolve, MakeEndpoint},
+        resolve::map_endpoint,
+    },
+    svc::{
+        layer,
+        stack::{FilterRequest, RequestFilter},
+        NewService,
+    },
+    Error,
+};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+type ResolveStack<F, R> = map_endpoint::Resolve<
+    endpoint::FromMetadata,
+    RecoverDefault<RequestFilter<F, ResolveService<R>>>,
+>;
+
+fn new_resolve<T, F, R>(filter: F, resolve: R) -> ResolveStack<F, R>
+where
+    T: Clone,
+    for<'t> &'t T: Into<std::net::SocketAddr>,
+    endpoint::FromMetadata: map_endpoint::MapEndpoint<T, Metadata>,
+    F: FilterRequest<T>,
+    R: Resolve<F::Request, Endpoint = Metadata>,
+    R::Future: Send + 'static,
+    R::Resolution: Send + 'static,
+{
+    map_endpoint::Resolve::new(
+        endpoint::FromMetadata,
+        RecoverDefault(RequestFilter::new(filter, resolve.into_service())),
+    )
```

Comment:
just a thought: it seems like, theoretically, we _could_ compose a lot of the resolver stuff using a stack rather than a bunch of nested constructors...maybe a future refactor?

Reply:
I think stacks are preferable when we have to compose many orthogonal concerns. In this case, I think it's simpler/clearer to just construct the services manually.